Adding f2py signature file to avoid using ctypes

gregstarr opened the pull request:
This PR is to fix apexpy on windows 10. I added a f2py signature file so that you don't have to use ctypes to use the `feldg` function in magfld.f. I added the signature file to the install script and removed `_cfa` from apex.py. I got all the tests to pass on Windows 10 and almost all of them to pass in Windows Subsystem for Linux. The failing ones I hope will pass in CI.

gregstarr commented:
Passed everything in 2.7-64 and 3.6-64, passed almost everything in 3.5-64. Couldn't compile for the others. Any ideas?

aburrell commented:
I suspect (but do not know for sure) that the .pyf file needs to be created on each system, and the systems that are passing are those most related to your system.  I think you're on the right track with upgrading the ctypes calls in the Apex class though.

gregstarr commented:
The same compilation / build errors occurred in the last commit so I don't think it has to do with fortranapex.pyf

https://ci.appveyor.com/project/aburrell/apexpy/builds/37593205

aburrell commented:
Also it passes on Travis, so that makes it more unlikely to be the cause.  

AACGMV2 had an appveyor problem recently and basically, it was down to fixing the testing environment (C was only available on limited platforms).  Could be something similar.

gregstarr commented:
Looking back, it doesn't look like the "32" builds ever passed. I see this error in most of them `cannot find -lpython36
collect2.exe: error: ld returned 1 exit status`

gregstarr commented:
Took a lot of trial and error (sorry about all the commits), but I got both 64 and 32 bit jobs working, as well as the PEP8 checker job. There are three tests failing on 64 bit python 3.5 related to the command line interface. I removed testing for python 3.4 because it looks like it is no longer supported by anaconda and I added 3.7. 

build inspired by: https://github.com/timcera/wdmtoolbox/blob/master/appveyor.yml

job early cancel: https://github.com/ogrisel/python-appveyor-demo/blob/master/appveyor.yml

aburrell commented:
I saw all the commits this morning and was cheering you on, thank you for figuring this out!  I propagated the changes you made to the appveyor template (the CI templates are located in `ci/templates` and are used along with `setup.cfg` to update `tox.ini`, `appveyor.yml`, and `.travis.yml`.  I also:
- added more recent python versions
- updated the appropriate docs.

Hopefully this will all pass and we can merge this in!

aburrell commented on appveyor.yml:
```diff
@@ -61,28 +62,32 @@ environment:
   - '%MINICONDA_HOME%\Scripts\conda update -q conda'
-  - '%MINICONDA_HOME%\Scripts\conda create -q -p C:\pythontest python=%PYTHON_VERSION% %INSTALL_LIBPYTHON% pytest numpy'
-  # remove file which conflicts with installation of package 'readme' on case-insensitive file systems
+  - '%MINICONDA_HOME%\Scripts\conda create -q -p C:\pythontest python=%PYTHON_VERSION% pytest numpy libpython'
```
Do we still need `libpython`?

gregstarr replied:
Yes I think we do, adding libpython back in was one of the last things I had to do to get it to build. Don't ask me why though

aburrell commented on ci/templates/appveyor.yml:
```diff
@@ -37,28 +24,32 @@ environment:
   - '%MINICONDA_HOME%\Scripts\conda update -q conda'
-  - '%MINICONDA_HOME%\Scripts\conda create -q -p C:\pythontest python=%PYTHON_VERSION% %INSTALL_LIBPYTHON% pytest numpy'
-  # remove file which conflicts with installation of package 'readme' on case-insensitive file systems
+  - '%MINICONDA_HOME%\Scripts\conda create -q -p C:\pythontest python=%PYTHON_VERSION% pytest numpy libpython'
```
If not, don't forget to remove it here first/as well.

aburrell commented on src/apexpy/apex.py:
```diff
@@ -13,13 +13,11 @@
 try:
     from . import fortranapex as fa
-except:
+except ImportError as e:
```
```suggestion
except ImportError as err:
```

aburrell commented on src/apexpy/apex.py:
```diff
@@ -13,13 +13,11 @@
-
-import ctypes
-
+    raise e
```
```suggestion
    raise err
```

aburrell replied:
To avoid single letter variable names that are not iterator indices.

gregstarr commented:
Thanks for taking a look. I notice that the same 64 bit python 3.5 tests are failing on master as well, so perhaps that could be tracked in a separate issue?

aburrell commented on src/apexpy/apex.py:
```diff
@@ -76,15 +74,15 @@ def __init__(self, date=None, refh=0, datafile=None, fortranlib=None):
                 self.year = helpers.toYearFraction(date)
-            except:
+            except AttributeError:
                 # failed so date is probably int/float, use directly
```
```suggestion
                # Failed while finding datetime attribute, so
                # date is probably an int or float; use directly
```

Just clarifying why the AttributeError is raised.

aburrell commented:
> Thanks for taking a look. I notice that the same 64 bit python 3.5 tests are failing on master as well, so perhaps that could be tracked in a separate issue?

It could, but numpy cut active support for python 3.5, so I don't think it's a problem worth spending a lot of energy on.  Current lowest supported python is 3.6: https://numpy.org/neps/nep-0029-deprecation_policy.html

aburrell commented:
The tests that are failing are tests that were fixed in the last pull request.  Any idea what's up?

gregstarr commented:
These are new failures, in the past the travis CI was only run on 2.7, 3.4, 3.5 and 3.6. Both 2.7 and 3.6 were passing before and are still passing now but the new ones 3.7, 3.8 and 3.9 are failing one test. It could be because of a change made in 3.7 on linux which messes with the test. I notice the test is trying to catch warnings and verify that there is exactly 1: 
`assert len(wmsg) == 1`
It is failing because `len(wmsg)` is 2. I will add a print statement so we can see whether a length of 2 is reasonable.

gregstarr commented:
There was a numpy deprecation warning which was causing 2 warnings instead of the expected 1. Should be all set now, let me know if there are any adjustments you want.

aburrell commented on tests/test_Apex.py:
```diff
@@ -393,6 +393,8 @@ def test_geo2apex_undefined_warning():
         apex2000 = Apex(date=2000, refh=10000)
         ret = apex2000.geo2apex(0, 0, 0)
 
+    print(wmsg)
+    print([w.message for w in wmsg])
```
```suggestion
```

aburrell reviewed (commented):
Fantastic, thanks for tracking that down.  I should have thought about deprecation warnings... ah well.  Only suggestion is to remove your debugging statements.  Should be good to go after that!
